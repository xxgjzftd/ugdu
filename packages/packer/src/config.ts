import { cwd } from 'process'

import { TaskOptions } from '@ugdu/processor'
import { mergeConfig } from 'vite'

import type { InlineConfig } from 'vite'
import type { Merge } from 'type-fest'

export interface UserConfig {
  extensions: string[]
  cwd?: string
  /**
   * @default 'dist'
   */
  dist?: string
  /**
   * @default 'assets'
   */
  assets?: string
  /**
   * @default '/'
   */
  base?: string
  apps: UserAppConfig[]
  routes?: Record<string, RoutesOption>
  vite?: InlineConfig
  meta: 'local' | `http${'s' | ''}://${string}/`
}

export interface UserAppConfig {
  name: string
  predicate?: (pathname: string) => boolean
  vite?: InlineConfig
  packages: ((packages: string[]) => string[]) | string[]
}

export interface RoutesOption {
  patterns: string | string[]
  base: `/${string}/` | '/'
  depth: number
  extends: RouteExtend[]
}

type RouteExtend = Partial<Pick<BaseRoute, 'path' | 'name' | 'depth'>> & Pick<BaseRoute, 'id'>

export interface BaseRoute {
  id: string
  path: string
  name: string
  depth: number
  component: string
  children?: BaseRoute[]
}

export type Config = Merge<Required<UserConfig>, { apps: Required<UserAppConfig>[] }>

declare module '@ugdu/processor' {
  interface Context {
    config: Config
  }
}

export interface SetConfigHooks {
  'get-config'(): UserConfig | void
}

export const setConfig = new TaskOptions<SetConfigHooks>(
  async function () {
    const config = await this.call('get-config', 'first')
    if (!config) {
      throw new Error(`The 'get-config' hook should have at least one hook function which returns a config.`)
    }
    config.cwd = config.cwd ?? cwd()
    config.dist = config.dist ?? 'dist'
    config.assets = config.assets ?? 'assets'
    config.base = config.base ?? '/'
    config.vite = mergeConfig(
      {
        base: config.base,
        build: {
          outDir: config.dist,
          assetsDir: config.assets,
          emptyOutDir: false,
          commonjsOptions: { esmExternals: true }
        }
      },
      config.vite ?? {}
    )
    config.routes = config.routes ?? {}
    config.apps.forEach(
      (app) => {
        app.predicate = () => true
        app.vite = {}
      }
    )
    this.manager.context.config = config as Config
  }
)
