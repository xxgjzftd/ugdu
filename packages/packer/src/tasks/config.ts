import { cwd } from 'process'

import { TaskOptions } from '@ugdu/processor'
import { mergeConfig, normalizePath } from 'vite'

import type { InlineConfig } from 'vite'
import type { Merge } from 'type-fest'

import type { PkgNode } from './project'

/**
 * @public
 */
export interface UserConfig {
  /**
   * In `local package`s which don't have `main` field, all files with extension in this config are considered to be a `module`.
   *
   * @remarks
   * There are two types of our source files.
   * One is `module`, the other is `source`.
   * `source`s are bundled into `module` when building.
   * `module`s import each other at runtime.
   * In packages which have `main` field, the corresponding file of `main` is considered to be a `module`,
   * and the other files are considered as `source`.
   * In packages which don't have `main` field, all files with extension in this config are considered to be a `module`.
   * Usually, this config will be `['js', 'ts', 'jsx', 'tsx']`.
   * According to the framework you use, you should specify the corresponding value.
   * Such as, for `vue`, you should add `vue` to this config.
   * Check {@link MetaModule} for more information about `module`.
   *
   */
  extensions: string[]
  /**
   * The current working directory.
   *
   * @remarks
   * This field affects following process
   *
   * - Where to find `local package` and `vendor package`
   *
   * - Where to write `dist` file
   *
   * - Where to find `all sources` and `changed sources`
   *
   * - In build process, the relative path to this field of a file is used to represent the `path` of that file
   *
   * @defaultValue `process.cwd()`
   */
  cwd?: string
  /**
   * The output directory.
   *
   * @defaultValue 'dist'
   */
  dist?: string
  /**
   * The directory to nest generated assets.
   *
   * @defaultValue 'assets'
   */
  assets?: string
  /**
   * Base public path when served in development or production.
   *
   * @defaultValue '/'
   */
  base?: string
  /**
   * The `app`s's configurations.
   *
   * @remarks
   * A project can contains multiple `app`s.
   * Different `app`s can be based on different fe frameworks.
   */
  apps: UserAppConfig[]
  /**
   * The vite config all `app`s should apply.
   *
   * @defaultValue \{\}
   */
  vite?: InlineConfig
  /**
   * Decide where to find previous build information.
   *
   * @remarks
   * `local` means that we should read the build information from where we should write this time.
   */
  meta: 'local' | `http${'s' | ''}://${string}/`
}

/**
 *
 * @remarks
 * The `app` can actually be thought of as the entry `package`.
 *
 * @public
 */
export interface UserAppConfig {
  /**
   * The name of `local package` which is the entry package of this app.
   */
  name: string
  /**
   * Whether to load and mount this `app`.
   */
  predicate?: (pathname: string) => boolean
  /**
   * The vite config which will be applied when building this `app`.
   */
  vite?: InlineConfig
  /**
   * The packages belonging to this `app`.
   */
  packages: ((packages: PkgNode[]) => string[]) | string[]
}

/**
 * The normalized config.
 *
 * @internal
 */
export type Config = Merge<Required<UserConfig>, { apps: Required<UserAppConfig>[] }>

/**
 * @public
 */
export interface SetConfigHooks {
  /**
   * A `first` type hook. Its first non null return value will be used as config after normalized.
   */
  'get-config'(): UserConfig | void
}

/**
 * Gets user's config from `get-config` hook and normalize it.
 * Then sets it to `context.config`.
 *
 * @public
 */
export const setConfig = new TaskOptions<SetConfigHooks>(
  async function () {
    const config = await this.call('get-config', 'first')
    if (!config) {
      throw new Error(`The 'get-config' hook should have at least one hook function which returns a config.`)
    }
    config.cwd = normalizePath(config.cwd ?? cwd())
    config.dist = config.dist ?? 'dist'
    config.assets = config.assets ?? 'assets'
    config.base = config.base ?? '/'
    config.vite = mergeConfig(
      {
        base: config.base,
        build: {
          outDir: config.dist,
          assetsDir: config.assets,
          emptyOutDir: false
        }
      },
      config.vite ?? {}
    )
    config.apps.forEach(
      (app) => {
        app.predicate = app.predicate ?? (() => true)
        app.vite = app.vite ?? {}
      }
    )
    this.manager.context.config = config as Config
  },
  ['get-config']
)
