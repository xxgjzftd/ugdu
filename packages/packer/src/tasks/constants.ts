import { TaskOptions } from '@ugdu/processor'
import { resolve } from 'path'

/**
 * @internal
 */
export const CONSTANTS = {
  /**
   * The filename of the file which store previous build info.
   */
  META_JSON: 'meta.json',
  PACKAGE_JSON: 'package.json',
  BINDING_NAME_SEP: '$ugdu',
  PACKAGE_NAME_SEP: '$ugdu',
  VERSIONED_VENDOR_SEP: '@',
  ROUTES: 'routes',
  ROUTES_INPUT: resolve('routes'),
  VENDOR: 'vendor',
  VENDOR_INPUT: resolve('vendor'),
  ROOT: 'root',
  INDEX: 'index'
} as const

/**
 * @internal
 */
export type CONSTANTS = typeof CONSTANTS

/**
 * Sets built-in CONSTANTS to `context.CONSTANTS`.
 *
 * @public
 */
export const setConstants = new TaskOptions(
  function () {
    this.manager.context.CONSTANTS = CONSTANTS
  }
)
