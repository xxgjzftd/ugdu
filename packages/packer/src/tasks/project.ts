import { readFile } from 'fs/promises'
import { resolve } from 'path'

import fg from 'fast-glob'
import dh from 'dependencies-hierarchy'
import fwp from '@pnpm/find-workspace-packages'
import axios from 'axios'
import { execa } from 'execa'
import { normalizePath } from 'vite'
import { parallel, series, TaskOptions } from '@ugdu/processor'

import { cached, clone, getDefault } from '../shared/utils'
import { setConstants } from './constants'
import { setConfig } from './config'
import { setUtils } from './utils'

import type { Promisable } from 'type-fest'
import type { PackageNode } from 'dependencies-hierarchy'
import type { AliasOptions } from 'vite'
import type { Context } from '@ugdu/processor'

/**
 * The information of your project.
 *
 * @public
 */
export interface Project {
  /**
   * The alias config used when building.
   */
  alias: AliasOptions
  /**
   * All `package`s including `local package`s and `vendor package`s in your project.
   */
  pkgs: PkgNode[]
  /**
   * The information of the building.
   */
  meta: {
    /**
     * The information of the previous building.
     */
    pre: Meta
    /**
     * The information of the current building.
     */
    cur: Meta
  }
  /**
   * The information of your source files.
   */
  sources: Sources
  /**
   * Used to generate the code of the `routes module`.
   *
   * @remarks
   * By default, this field will be generated according to your project structure.
   * Files in `src/pages` of `local package`s will be considered to be `page`s.
   * A `page` corresponds to a `BaseRoute`.
   * For its path, we need replace the `${local package path}/src/pages` with {@link Utils.getPkgId | `package` id} and then remove the file extension.
   * For dynamic routes, you should name your file within [].
   * For nested routes, you should put the children pages in the folder which has the same name with the parent page.
   * If there is a file named `index`, the generated path will remove it automaticly.
   * If there is a file whose name is the same as the {@link Utils.getPkgId | `package` id}, and it's located at the root of the `src/pages`, then it will be the parent page of the other pages in this package.
   * If there is a package whose id is `root`, we will remove the `root` from the generated path. You may need to put the pages which is no need to make a package in this package. Such as `/login`, `/welcom` etc.
   * 
   * @example
   * Suppose we have the following structure:
   * ```
   * - packages/
   *   - foo/
   *     - package.json
   *     - src/
   *       - pages/
   *         - xx.vue
   *         - yy.vue
   *         - xx/
   *           - zz.vue
   *   - bar/
   *     - package.json
   *     - src/
   *       - pages/
   *         - bar.vue
   *         - xx/
   *           - index.vue
   *           - [id].vue
   *   - root/
   *     - package.json
   *     - src/
   *       - pages/
   *         - index.vue
   *         - login.vue
   * ```
   * 
   * The generated `routes` module will be something like:
   * ```
   * [
   *   {
   *     path: '/foo/xx',
   *     component: 'packages/foo/src/pages/xx.vue',
   *     children: [
   *       {
   *         path: '/foo/xx/zz',
   *         component: 'packages/foo/src/pages/xx/zz.vue'
   *       }
   *     ]
   *   },
   *   {
   *     path: '/foo/yy',
   *     component: 'packages/foo/src/pages/yy.vue'
   *   },
   *   {
   *     path: '/bar',
   *     component: 'packages/bar/src/pages/bar.vue',
   *     children: [
   *       {
   *         path: '/bar/xx',
   *         component: 'packages/bar/src/pages/xx/index.vue'
   *       },
   *       {
   *         path: '/bar/xx/:id',
   *         component: 'packages/bar/src/pages/xx/[id].vue'
   *       }
   *     ]
   *   },
   *   {
   *     path: '/',
   *     component: 'packages/root/src/pages/index.vue'
   *   },
   *   {
   *     path: '/login',
   *     component: 'packages/root/src/pages/login.vue'
   *   }
   * ]
   * ```
   */
  routes: BaseRoute[]
}

/**
 * @public
 */
export interface PkgNode {
  name: string
  version: string
  /**
   * The `path` of the folder for this package.
   * Check {@link UserConfig.cwd} for what `path` is.
   *
   * @remarks
   * For `local package`s, the folder is the package's folder.
   * For `vendor package`s, the folder has a subfolder named 'node_modules' which contains the package's folder.
   */
  path: string
  /**
   * The absolute path of the folder for this package.
   */
  ap: string
  /**
   * Is this package a `local package`.
   */
  local: boolean
  /**
   * The `main` field of this package.
   */
  main?: string
  /**
   * Packages which depend on this package.
   */
  dependents: PkgNode[]
  /**
   * This package's dependencies.
   */
  dependencies: PkgNode[]
}

/**
 * @internal
 */
export type LocalPkgToDepsMap = Map<PkgNode, PackageNode[]>

/**
 * The build information.
 *
 * @public
 */
export interface Meta {
  /**
   * All `module`s in this project.
   */
  modules: MetaModule[]
  hash?: string
  version?: string
}

/**
 * The information of a `module`.
 *
 * @remarks
 * There are three types `module` in our project.
 *
 * First is `local module`.
 * `local module` comes from `local package`.
 * For a `local package` which have `main` field, the corresponding file of `main` is a `local module`, and the other files will be bundled into this module.
 * For other `local package`s, all files with extension in {@link UserConfig.extensions} are considered to be a `local module`.
 *
 * Second is `vendor module`.
 * `vendor module` comes from `vendor package`.
 * Only `vendor package`s imported by multiple `module`s or `local module` have a corresponding `vendor module`.
 * The other `vendor package`s will be bundled into the `vendor module` which import them.
 *
 * Last is `routes module`.
 * `routes module` is generated according to the structure of your project.
 *
 * @public
 */
export interface MetaModule {
  /**
   * The module name.
   * There are three types of module names.
   *
   * First is `local module` name.
   * Such as `@pkg/components`, `@pkg/sale/src/pages/order/index.vue`.
   *
   * Second is `routes module` name(i.e.`routes`).
   *
   * Last is `vendor module` name.
   * Such as `lodash@4.17.21`
   * Here the `vendor module` name actually is versioned vendor package name.
   */
  id: string
  /**
   * The generated js file url.
   */
  js: string
  /**
   * The generated css file url. This field may not exist if this module doesn't have any css.
   */
  css?: string
  /**
   * The static resources this module depends. Such as css, img etc.
   * Only local modules have this field when it does depend any static resources.
   */
  sources?: string[]
  /**
   * All `public package name`s this module may depends.
   * A vendor module will not be rebuilt unless one of it's externals and exports changed.
   * Only vendor modules have this field.
   */
  externals?: string[]
  /**
   * The import info of this module include from which package and import what.
   */
  imports: MetaModuleImport[]
  /**
   * The exported variable names.
   * Used to check if there is any error in the local modules build.
   * Only local modules have this field.
   */
  exports?: string[]
}

/**
 * @public
 */
export interface MetaModuleImport {
  /**
   * The module name.
   */
  id: string
  /**
   * The public package name.
   */
  name: string
  /**
   * The bindings that the parent module import.
   */
  bindings: string[]
}

/**
 * @public
 */
export interface Sources {
  /**
   * All `path`s of your source files.
   */
  all: string[]
  /**
   * The `path` and status of your source files which changed since last build.
   */
  changed: ChangedSource[]
}

/**
 * @public
 */
export interface ChangedSource {
  /**
   * 'A' means the file is added.
   * 'M' means the file is modified.
   * 'D' means the file is deleted.
   */
  status: 'A' | 'M' | 'D'
  /**
   * The `path` of the source file.
   */
  path: string
}

/**
 * Used to generate the code of the `routes module`.
 *
 * @public
 */
export interface BaseRoute {
  /**
   * The file path relative to {@link UserConfig.cwd}.
   */
  id: string
  path: string
  name: string
  /**
   * The file path relative to {@link UserConfig.cwd}. In the generated code, this field will be something like '() =\> import("/path/to/this/component")'.
   */
  component: string
  children?: BaseRoute[]
}

const getLocalPkgs = async (cwd: string) =>
  (await getDefault(fwp)(cwd)).slice(1).map(
    (pkg) => {
      const {
        manifest: { name, main, version },
        dir
      } = pkg
      if (!name) throw new Error(`The package at '${pkg.dir}' doesn't specified the 'name' field.`)
      if (!version) throw new Error(`The package at '${pkg.dir}' doesn't specified the 'version' field.`)
      const ap = normalizePath(dir)
      return {
        name,
        version,
        path: ap.slice(cwd.length + 1),
        ap,
        local: true,
        main,
        dependents: [],
        dependencies: []
      } as PkgNode
    }
  )

const getLocalPkgToDepsMap = async (localPkgs: PkgNode[], cwd: string) =>
  getDefault(dh)(
    localPkgs.map((lp) => lp.ap),
    {
      depth: Infinity,
      include: { dependencies: true, devDependencies: false, optionalDependencies: false },
      lockfileDir: cwd
    }
  ).then(
    (pd2dhm) => {
      const lp2dm: LocalPkgToDepsMap = new Map()
      for (const pd of Object.keys(pd2dhm)) {
        lp2dm.set(localPkgs.find((lp) => lp.ap === pd)!, pd2dhm[pd].dependencies!)
      }
      return lp2dm
    }
  )

/**
 * @internal
 */
export const getPkgId = cached((lpn) => lpn.replace(/.+\//, ''))

/**
 * @internal
 */
export const getAliasKey = cached((lpn) => `@${getPkgId(lpn)}`)

const getAlias = (localPkgs: PkgNode[], context: Context) => {
  const {
    utils: { getPkgId }
  } = context
  const alias: AliasOptions = {}
  localPkgs.forEach((lp) => (alias[`@${getPkgId(lp.name)}`] = resolve(lp.ap, 'src')))
  if (Object.keys(alias).length < localPkgs.length) {
    throw new Error(`There are duplicate pkg id in local packages.`)
  }
  return alias
}

/**
 * @internal
 */
export const getPkgs = (localPkgToDepsMap: LocalPkgToDepsMap, cwd: string) => {
  const localPkgs = [...localPkgToDepsMap.keys()]
  const pkgs = [...localPkgs]
  const traverse = (deps: PackageNode[], dependent: PkgNode, pp: string) => {
    deps.forEach(
      (dep) => {
        let pkg = findPkgFromDep(dep)
        if (!pkg) {
          const ap = normalizePath(dep.path).replace(pp, cwd)
          pkg = {
            name: dep.name,
            version: dep.version,
            path: ap.slice(cwd.length + 1),
            ap,
            local: false,
            dependencies: [],
            dependents: []
          }
          pkgs.push(pkg)
          dep.dependencies && traverse(dep.dependencies, pkg, pp)
        }
        pkg.dependents.push(dependent)
        dependent.dependencies.push(pkg)
      }
    )
  }
  const findPkgFromDep = (dep: PackageNode) =>
    pkgs.find((pkg) => pkg.name === dep.name && (isLocalPkg(pkg) || pkg.version === dep.version))
  const isLocalPkg = (n: PkgNode) => localPkgs.find((lp) => lp.name === n.name)
  for (const localPkg of localPkgToDepsMap.keys()) {
    traverse(localPkgToDepsMap.get(localPkg)!, localPkg, localPkg.ap)
  }
  return pkgs
}

const getAllPkgs = async (localPkgs: PkgNode[], cwd: string) => {
  const localPkgToDepsMap = await getLocalPkgToDepsMap(localPkgs, cwd)
  const pkgs = getPkgs(localPkgToDepsMap, cwd)
  return pkgs
}

const getPreMeta = async (context: Context) => {
  let meta: Meta
  const {
    CONSTANTS: { META_JSON },
    config,
    config: { cwd, dist }
  } = context
  try {
    if (config.meta === 'local') {
      meta = JSON.parse(await readFile(resolve(cwd, dist, META_JSON), 'utf-8'))
    } else {
      meta = await axios.get(`${config.meta}${META_JSON}`).then((res) => res.data)
    }
  } catch (error) {
    meta = { modules: [] }
  }
  // meta.json maybe empty
  meta.modules = meta.modules || []
  return meta
}

const getCurMeta = async (context: Context) => {
  const {
    project: {
      meta: { pre }
    },
    config: { cwd }
  } = context
  const cur: Meta = { modules: [] }
  const { stdout } = await execa('git', ['rev-parse', '--short', 'HEAD'], { cwd })
  cur.hash = stdout
  cur.version = VERSION
  cur.modules = clone(pre.modules.filter((m) => !m.externals))
  return cur
}

const getSources = async (context: Context) => {
  const {
    config: { cwd },
    project: {
      pkgs,
      meta: { pre }
    }
  } = context
  const all = await fg(
    pkgs.filter((pkg) => pkg.local).map((lp) => lp.path + '/**'),
    {
      cwd,
      ignore: ['**/node_modules/**']
    }
  )
  let changed: ChangedSource[] = []
  if (pre.hash && pre.version === VERSION) {
    const { stdout } = await execa('git', ['diff', pre.hash, 'HEAD', '--name-status'], { cwd })
    changed = stdout
      .split('\n')
      .map(
        (info) => {
          const [status, path] = info.split('\t')
          return { status, path } as ChangedSource
        }
      )
      .filter(({ path }) => all.includes(path))
  } else {
    changed = all.map(
      (path) => {
        return { status: 'A', path }
      }
    )
  }
  return { all, changed }
}

const getRoutes = async (context: Context) => {
  const {
    CONSTANTS: { ROOT, INDEX },
    utils: { appendSlash, getLocalPkgs, getPkgId }
  } = context
  const routes: BaseRoute[] = []

  await Promise.all(
    getLocalPkgs()
      .filter((lp) => !lp.main)
      .map(
        async (lp) => {
          const pp = `${lp.path}/src/pages`
          const pi = getPkgId(lp.name)
          const paths = await fg(`${pp}/**/*`)
          const brs = paths.map(
            (path) => {
              const br: BaseRoute = {
                id: path,
                path: path
                  .replace(pp, '/' + pi)
                  .replace(/\.[^\.]+$/, '')
                  .replace(/(?<=\/)\[(.+?)\](?=(\/|$))/g, ':$1')
                  .replace(new RegExp(`^(/${pi})\\1$`), '$1'),
                name: '',
                component: path
              }
              return br
            }
          )
          const insert = (target: BaseRoute, list: BaseRoute[]) => {
            for (let i = 0; i < list.length; i++) {
              const current = list[i]
              if (target.path.startsWith(appendSlash(current.path))) {
                current.children = current.children || []
                insert(target, current.children)
                return
              } else if (current.path.startsWith(appendSlash(target.path))) {
                target.children = []
                list.splice(i, 1, target)
                insert(current, target.children)
                return
              }
            }
            list.push(target)
          }
          const sub: BaseRoute[] = []
          brs.forEach((br) => insert(br, sub))
          if (pi === ROOT) {
            brs.forEach((br) => (br.path = br.path.replace(new RegExp(`^/${ROOT}`), '')))
          }
          brs.forEach(
            (br) => {
              br.path = br.path.replace(new RegExp(`/${INDEX}$`), '')
              br.name = br.path.slice(1).replace(/\//g, '-')
            }
          )
          routes.push(...sub)
        }
      )
  )

  return routes
}

/**
 * @public
 */
export interface SetProjectHooks {
  /**
   * A `first` type hook. Its first non null return value will be used as `local package`s.
   */
  'get-local-packages'(cwd: string): Promisable<PkgNode[]>
  /**
   * A `first` type hook. Its first non null return value will be used as {@link Project.alias}.
   */
  'get-alias'(localPkgs: PkgNode[], context: Context): Promisable<AliasOptions>
  /**
   * A `first` type hook. Its first non null return value will be used as {@link Project.pkgs}.
   */
  'get-all-packages'(localPkgs: PkgNode[], cwd: string): Promisable<PkgNode[]>
  /**
   * A `first` type hook. Its first non null return value will be used as {@link Project.meta.pre}.
   */
  'get-previous-meta'(context: Context): Promisable<Meta>
  /**
   * A `first` type hook. Its first non null return value will be used as {@link Project.meta.cur}.
   */
  'get-current-meta'(context: Context): Promisable<Meta>
  /**
   * A `first` type hook. Its first non null return value will be used as {@link Project.sources}.
   */
  'get-sources'(context: Context): Promisable<Sources>
  /**
   * A `first` type hook. Its first non null return value will be used as {@link Project.routes}.
   */
  'get-routes'(context: Context): Promisable<BaseRoute[]>
}

/**
 * @public
 */
export const setProject = series(
  parallel(setConstants, setConfig, setUtils),
  new TaskOptions<SetProjectHooks>(
    async function () {
      const {
        manager: {
          context,
          context: {
            config: { cwd }
          }
        }
      } = this
      const project = { meta: {} } as Project
      context.project = project
      const localPkgs = await this.call('get-local-packages', 'first', cwd)
      project.alias = await this.call('get-alias', 'first', localPkgs, context)
      project.pkgs = await this.call('get-all-packages', 'first', localPkgs, cwd)
      project.meta.pre = await this.call('get-previous-meta', 'first', context)
      project.meta.cur = await this.call('get-current-meta', 'first', context)
      project.sources = await this.call('get-sources', 'first', context)
      project.routes = await this.call('get-routes', 'first', context)
    },
    [
      'get-local-packages',
      'get-alias',
      'get-all-packages',
      'get-previous-meta',
      'get-current-meta',
      'get-sources',
      'get-routes'
    ],
    {
      'get-local-packages': getLocalPkgs,
      'get-alias': getAlias,
      'get-all-packages': getAllPkgs,
      'get-previous-meta': getPreMeta,
      'get-current-meta': getCurMeta,
      'get-sources': getSources,
      'get-routes': getRoutes
    }
  )
)
