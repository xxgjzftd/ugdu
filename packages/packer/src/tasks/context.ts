import { parallel } from '@ugdu/processor'

import { setConfig } from './config'
import { setConstants } from './constants'
import { setProject } from './project'
import { setUtils } from './utils'

export const setContext = parallel(setConstants, setConfig, setProject, setUtils)
