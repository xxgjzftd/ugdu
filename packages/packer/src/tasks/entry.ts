import { series, TaskOptions } from '@ugdu/processor'
import { build, mergeConfig } from 'vite'

import { setContext } from './context'
import { entry } from '../plugins/entry'

/**
 * Build the entry of the project.
 *
 * @remarks
 * It creates the `importmap` and startup script of {@link https://github.com/xxgjzftd/ugdu/blob/main/packages/runtime/README.md | @ugdu/runtime}
 * according to the `module`s of this project. Then inject them to `index.html`.
 *
 * @public
 */
export const buildEntry = series(
  setContext,
  new TaskOptions(
    async function () {
      const {
        manager: { context }
      } = this
      build(
        mergeConfig(
          {
            plugins: [entry(context)]
          },
          context.config.vite
        )
      )
    }
  )
)
