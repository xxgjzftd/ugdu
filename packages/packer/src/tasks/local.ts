import { join } from 'path'

import { series, TaskOptions } from '@ugdu/processor'
import { build, mergeConfig } from 'vite'

import { setContext } from './context'
import { local } from '../plugins/local'
import { meta } from '../plugins/meta'
import { clone } from '../shared/utils'

import type { InlineConfig } from 'vite'
import type { Promisable } from 'type-fest'
import type { Context } from '@ugdu/processor'

declare module './project' {
  interface Project {
    mn2bm: ModuleNameToBindingsMap
  }
}

type ModuleNameToBindingsMap = { pre: Record<string, string[]>; cur: Record<string, string[]> }

const getModuleNameToBindingsMap = (context: Context) => {
  const {
    project: { meta }
  } = context
  const mn2bm: ModuleNameToBindingsMap = { pre: {}, cur: {} }
  const types = ['pre', 'cur'] as const
  types.forEach(
    (type) => {
      const map = mn2bm[type]
      meta[type].modules.forEach(
        (m) => {
          m.imports.forEach(
            (i) => {
              const mn = i.id
              const bindings = map[mn] || (map[mn] = [])
              i.bindings.forEach((binding) => bindings.push(binding))
            }
          )
        }
      )
      Object.keys(map).forEach((mn) => (map[mn] = [...new Set(map[mn])].sort()))
    }
  )

  return mn2bm
}

/**
 * @public
 */
export interface BuildLocalModulesHooks {
  /**
   * A `parallel` type hook. It will be invoked once a `local module` need be built.
   *
   * @remarks
   * Note: This hook will be triggered by the change of the `module`'s own file or its `source` file.
   *
   * @param lmn - `local module` name
   * @param context - {@link @ugdu/processor#Context}
   */
  'build-local-module'(lmn: string, context: Context): Promisable<void>
}

/**
 * @internal
 */
export const buildLocalModule = async function (lmn: string, context: Context) {
  const {
    config,
    config: { apps, assets },
    project: { alias },
    utils: { resolve, getLocalModulePath, getLocalModuleExternal, getPkgName }
  } = context
  const pn = getPkgName(lmn)
  const app = apps.find((app) => (app.packages as string[]).includes(pn))
  if (!app) {
    throw new Error(`'${pn}' doesn't have corresponding app.`)
  }
  const dc: InlineConfig = {
    publicDir: false,
    resolve: {
      alias
    },
    build: {
      rollupOptions: {
        input: resolve(getLocalModulePath(lmn)),
        output: {
          entryFileNames: join(assets, pn, '[name].[hash].js'),
          chunkFileNames: join(assets, pn, '[name].[hash].js'),
          assetFileNames: join(assets, pn, '[name].[hash][extname]'),
          format: 'es'
        },
        preserveEntrySignatures: 'allow-extension',
        external: getLocalModuleExternal(lmn)
      }
    },
    plugins: [local(lmn, context), meta(lmn, context)]
  }

  await build(mergeConfig(dc, mergeConfig(config.vite, app.vite)))
}

/**
 * Builds `local module`s.
 *
 * @remarks
 * `local module` is the `module` of `local package`s.
 * Check {@link MetaModule} for more information about `module`.
 *
 * @public
 */
export const buildLocalModules = series(
  setContext,
  new TaskOptions<BuildLocalModulesHooks>(
    async function build () {
      const {
        manager: {
          context,
          context: {
            config: { apps },
            utils: {
              getLocalPkgs,
              getLocalModuleName,
              addMetaModule,
              getPkgFromModuleName,
              getModuleNameFromPublicPkgName
            },
            project: {
              sources: { all, changed },
              meta: { pre, cur }
            }
          }
        }
      } = this

      apps.forEach(
        (app) => {
          if (typeof app.packages === 'function') {
            app.packages = app.packages(getLocalPkgs())
          }
        }
      )

      all.forEach(
        (path) => {
          const lmn = getLocalModuleName(path)
          const pmm = pre.modules.find((m) => m.id === lmn)
          if (lmn && pmm) {
            const cloned = clone(pmm)
            cloned.imports.forEach(
              (mmi) => {
                mmi.id = getModuleNameFromPublicPkgName(getPkgFromModuleName(lmn), mmi.name)
              }
            )
            addMetaModule(cloned)
          }
        }
      )

      const pending = new Set<string>()

      changed.forEach(
        ({ path }) => {
          const lmn = getLocalModuleName(path)
          if (lmn) {
            pending.add(lmn)
          } else {
            cur.modules.filter((m) => m.exports && m.sources?.includes(path)).forEach((m) => pending.add(m.id))
          }
        }
      )

      await Promise.all([...pending].map(async (lmn) => this.call('build-local-module', 'parallel', lmn, context)))
    },
    ['build-local-module'],
    {
      'build-local-module': buildLocalModule
    }
  ),
  new TaskOptions(
    async function setModuleNameToBindingsMap () {
      const {
        manager: {
          context,
          context: { project }
        }
      } = this
      project.mn2bm = getModuleNameToBindingsMap(context)
    }
  ),
  new TaskOptions(
    async function qa () {
      const {
        manager: {
          context: {
            project: { mn2bm, meta },
            utils: { isLocalModule, getLocalPkgFromName }
          }
        }
      } = this
      const lms = meta.cur.modules.filter((m) => isLocalModule(m.id))
      lms.forEach(
        (m) => {
          const mn = m.id
          const missing = mn2bm.cur[mn]?.find((b, index) => !m.exports!.includes(b, index))
          if (missing) {
            const dependents = meta.cur.modules
              .filter((m) => m.imports.find((i) => i.id === mn && i.bindings.includes(missing)))
              .map((m) => m.id)
            throw new Error(
              `Module '${mn}' no longer exports '${missing}' in this build.` +
                `But ${dependents.join(',')} still import it.`
            )
          }

          m.imports.forEach(
            (i) => {
              if (isLocalModule(i.id) && !lms.find((m) => m.id === i.id)) {
                throw new Error(
                  `Module '${i.id}' is imported by '${mn}' but it is not a local module name.` +
                    `You may need to import from ${getLocalPkgFromName(i.id)} instead of ${i.id}.`
                )
              }
            }
          )
        }
      )
    }
  )
)
