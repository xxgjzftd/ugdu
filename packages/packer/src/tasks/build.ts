import { parallel, series, TaskOptions } from '@ugdu/processor'

import { setContext } from './context'
import { buildLocalModules } from './local'
import { buildRoutesModules } from './routes'
import { buildVendorModules } from './vendor'
import { buildEntryModule } from './entry'
import { write } from './write'

/**
 * The `build` task options.
 *
 * @remarks
 * This task contains the following subtasks
 * - set `context.building` to `true`
 * - set `context`
 * - build `local module` and `routes module` in parallel
 * - build `vendor module`
 * - build `entry module`
 * - write the info of this build to disk
 *
 * @public
 */
export const build = series(
  new TaskOptions(
    function setBuilding () {
      this.manager.context.building = true
    }
  ),
  setContext,
  parallel(buildLocalModules, buildRoutesModules),
  buildVendorModules,
  buildEntryModule,
  write
)
