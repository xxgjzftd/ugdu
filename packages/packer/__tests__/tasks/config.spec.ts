import { expect, it } from 'vitest'
import { Processor } from '@ugdu/processor'

import { setConfig } from 'src/tasks/config'

import type { UserConfig } from 'src'

it('should throw an error if there is not a hook function which hooks into the get-config hook and returns a config', async () => {
  const task = new Processor().task(setConfig)
  await expect(task.run()).rejects.toThrow()
})

it("should get user's config from get-config hook and normalizes it, then set it to context.config", async () => {
  const task = new Processor().task(setConfig)
  const uc: UserConfig = {
    cwd: '/path/to/project',
    apps: [{ name: '@xx/container', packages: (lp) => lp.map((p) => p.name) }],
    extensions: ['ts', 'tsx'],
    meta: 'local'
  }
  task.hook('get-config', () => uc)
  await task.run()
  expect(task.manager.context.config).toMatchSnapshot()
})
