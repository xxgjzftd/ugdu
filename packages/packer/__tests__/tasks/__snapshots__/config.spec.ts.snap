// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should get user's config from get-config hook and normalizes it, then set it to context.config 1`] = `
{
  "apps": [
    {
      "name": "@xx/container",
      "packages": [Function],
      "predicate": [Function],
      "vite": {},
    },
  ],
  "assets": "assets",
  "base": "/",
  "cwd": "/path/to/project",
  "dist": "dist",
  "extensions": [
    "ts",
    "tsx",
  ],
  "meta": "local",
  "vite": {
    "base": "/",
    "build": {
      "assetsDir": "assets",
      "emptyOutDir": false,
      "outDir": "dist",
    },
  },
}
`;
